### PR TITLE
Added better variable handling of DSPO deployed manifest names

### PR DESCRIPTION
**Pull request by hbelmiro**
## The issue resolved by this Pull Request:
Resolves https://github.com/opendatahub-io/data-science-pipelines-operator/issues/165

## Description of your changes:
This PR adds better variable handling of DSPO deployed manifest names and improves tests by ensuring readiness is being properly handled

## Testing instructions
`make unittest`

## Checklist
- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by dsp-developers** [comment]
A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-408`
An OCP cluster where you are logged in as cluster admin is required.

To use this image run the following:

```bash
cd $(mktemp -d)
git clone git@github.com:opendatahub-io/data-science-pipelines-operator.git
cd data-science-pipelines-operator/
git fetch origin pull/408/head
git checkout -b pullrequest ec37818122b85402e0d6716e630fcd35f867758a
make deploy IMG="quay.io/opendatahub/data-science-pipelines-operator:pr-408"
```

More instructions [here](https://github.com/opendatahub-io/data-science-pipelines-operator#deploy-dsp-instance) on how to deploy and test a Data Science Pipelines Application.

**Review by rimolive — COMMENTED** [review]
/lgtm

**Review by DharmitD — APPROVED** [review]
Tested locally to make sure the changes work as expected.

/Approve

File: controllers/apiserver.go [hunk 1]
```diff
@@ -60,14 +72,28 @@ func (r *DSPAReconciler) ReconcileAPIServer(ctx context.Context, dsp *dspav1alph
...
 
 	for _, template := range apiServerTemplates {
-		err := r.Apply(dsp, params, template)
+		tmplManifest, err := r.manifest(template, params)
+		if err != nil {
+			return err
+		}
+
+		if template == serverDeployment {
+			r.addComponentName(apiServerComponentName, tmplManifest.Resources()[0].GetName())
```
Comment (HumairAK) [inline controllers/apiserver.go:81]: This seems like a lot of work to fetch a Deployment name. Would it not be simpler to specify in the `params` struct 3 new fields: 

```
APIServerDeploymentName
PersistentAgentDeploymentName
ScheduledWorkflowDeploymentName
```

Then utilize these here: [1](https://github.com/opendatahub-io/data-science-pipelines-operator/blob/main/config/internal/apiserver/deployment.yaml.tmpl#L4), [2](https://github.com/opendatahub-io/data-science-pipelines-operator/blob/main/config/internal/persistence-agent/deployment.yaml.tmpl#L4), [3](https://github.com/opendatahub-io/data-science-pipelines-operator/blob/main/config/internal/scheduled-workflow/deployment.yaml.tmpl#L4)

And let params struct be the source of truth for these names, accessible throughout the codebase, as one of the first things we do is generate this struct before doing anything else.

Comment (hbelmiro) [reply controllers/apiserver.go]: Makes total sense @HumairAK.
I pushed the changes. Can you have a second look, please?

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-408`

File: config/internal/apiserver/deployment.yaml.tmpl [hunk 2]
```diff
@@ -201,7 +201,7 @@ spec:
             - mountPath: /etc/tls/private
               name: proxy-tls
         {{ end }}
-      serviceAccountName: ds-pipeline-{{.Name}}
+      serviceAccountName: {{.APIServerDeploymentName}}
```
Comment (HumairAK) [inline config/internal/apiserver/deployment.yaml.tmpl:204]: this is tied to: https://github.com/opendatahub-io/data-science-pipelines-operator/blob/main/config/internal/apiserver/sa_ds-pipeline.yaml.tmpl#L4 

so if we change the service account name here, we need to do the same for the SA metadata.name
at which point, it is probably not appropriate to call it `APIServerDeploymentName` maybe something like `APIServerDefaultResourceName`. Same for persistent agent/swf.

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-408`

**Comment by HumairAK** [comment]
awesome thank you @hbelmiro 
/lgtm
/approve
/unhold

**Pull request merged**